keps/sig-cloud-provider/20190722-ccm-instance-metadata.md

```diff
@@ -0,0 +1,242 @@
+- For different Azure APIs, there’re also additional rate limits based on different durations. For example, there are 3Min and 30Min read limits for VMSS APIs (the numbers below are only for reference since they’re not officially [documented](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-request-limits)):
+  - Microsoft.Compute/HighCostGetVMScaleSet3Min;200
+  - Microsoft.Compute/HighCostGetVMScaleSet30Min;1000
+  - Microsoft.Compute/VMScaleSetVMViews3Min;5000
+
+Based on those rate limits, getting node’s information for a 5000 cluster may need hours.  Things would be much worse for multiple clusters in the same tenant and subscription.
+
+So the proposal aims to add IMDS support back with CCM, so that the kubernetes cluster could still be scaled to a large number of nodes.
+
+### Goals
+
+- Allow nodes to be initialized from IMDS.
+- Allow nodes to reconcile the node addresses from IMDS.
+
+### Non-Goals
+
+- Authentication and authorization for each provider implementations.
+- API throttling [issue](https://github.com/kubernetes/kubernetes/issues/60646) on route controller.
+
+## Proposal
+
+Same as kube-controller-manager, the [cloud provider interfaces](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L43) could be split into two parts:
+
+- Instance-level interfaces: Instances and Zones
+- Control-plane interfaces, e.g. LoadBalancer and Routes
+
+The control-plane interfaces are still kept in CCM (who’s name is cloud-controller-manager), and deployed on masters. For instance-level interfaces, a new daemonsets would be introduced and implement instance-level interfaces (who’s name is cloud-node-manager).
+
+With these changes, the whole node initialization workflow would be:
+
+- Kubelet specifying `--cloud-provider=external` will add a taint `node.cloudprovider.kubernetes.io/uninitialized` with an effect NoSchedule during initialization.
+- cloud-node-manager would initialize the node again with `Instances` and `Zones`.
+- cloud-controller-manager then take care of the rest things, e.g. configure the Routes and LoadBalancer for the node.
+
+Considering some providers may not require IMDS, cloud-node-manager cloud be enabled optionally by a new option --enable-node-controller on cloud-controller-manager.
+
+## Alternatives
+
+Since there are already a lot of plugins in Kubelet, e.g. CNI, CRI, and CSI. An alternative way is introducing another cloud-provider plugin, e.g. Cloud Provider Interface (CPI).
+
+When Kubelet starts, the cloud provider plugin may register itself into Kubelet, and then Kubelet invokes cloud provider plugin to initialize the node.
+
+One problem is the deployment of those plugins. If daemonsets is used to deploy those cloud provider plugins, then they should be schedulable before kubelet fully initialized the nodes. That means Kubelet may need to initialize itself two times:
```

Comment:
Same problem applies on `cloud-node-manager` executable, right? the assumption here is they are deployed as static pods, or cloud provider specific deployment technique?

Reply:
They are different:

* With plugin, kubelet need to invoke the plugins to complete the node registry. That means Kubelet needs to wait for plugin operations.
* With addon , kubelet won't invoke the cloud-node-manager. So, cloud-node-manager won't block any functionality from kubelet.